### PR TITLE
 Add deadline

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses #46

### CODE REVIEW

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 1]
```diff
@@ -61,8 +65,7 @@ protected override string PeerCore
             }
         }
 
-        // TODO(JunTaoLuo, JamesNK): implement this
-        protected override DateTime DeadlineCore => throw new NotImplementedException();
+        protected override DateTime DeadlineCore => _deadline;
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Note that the deadline support means two things:
- exposing the deadline to the user code via server call context
- the framework itself should watch the deadline and terminate the http stream if deadline is reached. Also, the  serverCallContext.CancellationToken should be automatically cancelled by the framework when the deadline is reached (another case when the cancellationToken should be triggered is when the client cancels).

File: testassets/FunctionalTestsWebsite/Greeter.cs [hunk 2]
```diff
@@ -37,7 +38,38 @@ public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContex
...
+    {
+        context.WriteResponseHeadersAsync(null);
+
+        return SayHellosCore(request, responseStream);
+    }
+
+    public override async Task SayHellosDeadline(HelloRequest request, IServerStreamWriter<HelloReply> responseStream, ServerCallContext context)
+    {
+        var i = 0;
+        while (DateTime.UtcNow < context.Deadline)
```
Comment (jtattermusch) [inline testassets/FunctionalTestsWebsite/Greeter.cs]: rather than watching the current time, we should probably watch the context.CancellationToken (and framework should cancel it when the deadline is crossed).

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 1 above)
Comment (jtattermusch) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: I think the right way to terminate a stream on the server side if deadline is reached is to send RST_STREAM (and not to send a status "DeadlineExceeded"), but I will have to confirm that. AFAIK the "DeadlineExceeded" status code is generated on the client side - with gRPC, both client and server are watching the deadline and both parties cancel if it's reached.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 3]
```diff
@@ -130,5 +133,51 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
...
+        {
+            const long TicksPerMicrosecond = 10;
+            const long NanosecondsPerTick = 100;
+
+            if (!HttpContext.Request.Headers.TryGetValue(GrpcProtocolConstants.TimeoutHeader, out var values))
+            {
+                return DateTime.MaxValue;
+            }
+
+            if (values.Count == 1)
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: nit: maybe the logic for parsing the timeout header shouldn't be directly in the ServerCallContext?

**Comment by AlgorithmsAreCool** [comment]
I'm curious if there is any concern about the system clock not being a stable enough time source for timeout logic. 

What if the clock syncs and jumps a minute into the past or future? Is that scenario rare enough to not worry about?

**Comment by JamesNK** [comment]
re: https://github.com/grpc/grpc-dotnet/pull/73#discussion_r254637889

The cancellation token source won't be effected by a changing system clock.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 4]
```diff
@@ -130,5 +133,51 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
 
             return HttpContext.Response.Body.FlushAsync();
         }
+
+        private DateTime GetDeadline()
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Curious where the logic to parse the deadline was specified? Was there a document on this I missed?

File: test/FunctionalTests/DeadlineTests.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+using Grpc.AspNetCore.Server.Internal;
+using NUnit.Framework;
+
+namespace Grpc.AspNetCore.FunctionalTests
+{
+    [TestFixture]
+    public class DeadlineTests : FunctionalTestBase
+    {
+        [Test]
+        public async Task WriteUntilDeadline_SuccessResponsesStreamed()
```
Comment (JunTaoLuo) [inline test/FunctionalTests/DeadlineTests.cs]: Add a test for the negative case? where the Deadline is exceeded?

File: test/FunctionalTests/Infrastructure/PipeReaderFixStream.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+#endregion
+
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.FunctionalTests.Infrastructure
+{
+    // Stream to work around StreamPipeReader issue
+    // https://github.com/aspnet/AspNetCore/issues/7329
```
Comment (JunTaoLuo) [inline test/FunctionalTests/Infrastructure/PipeReaderFixStream.cs]: 😢 Add a reminder (TODO or issue) to remove this workaround

File: test/FunctionalTests/ServerStreamingMethodTests.cs [hunk 7]
```diff
@@ -114,12 +117,55 @@ public async Task WriteResponseHeadersAsyncCore_FlushesHeadersToClient()
...
             Assert.False(goodbyeTask.IsCompleted);
             Assert.AreEqual("Goodbye World!", (await goodbyeTask.DefaultTimeout()).Message);
 
-            var finishedTask = MessageHelpers.AssertReadMessageStreamAsync<HelloReply>(responseStream);
+            var finishedTask = MessageHelpers.AssertReadStreamMessageAsync<HelloReply>(pipeReader);
             Assert.IsNull(await finishedTask.DefaultTimeout());
         }
+
+        [Test]
+        public async Task Buffering_SuccessResponsesStreamed()
```
Comment (JunTaoLuo) [inline test/FunctionalTests/ServerStreamingMethodTests.cs]: Is this a little redundant? ServerStreaming and DuplexStreaming mechanisms are both using a `IServerStreamWriter`

File: testassets/FunctionalTestsWebsite/Greeter.cs (hunk 2 above)
Comment (JamesNK) [reply testassets/FunctionalTestsWebsite/Greeter.cs]: I've added a test for that as well.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: I've made the CancellationToken use the deadline. Let me know what cancellation should do to the http stream.

**Comment by JamesNK** [comment]
🆙 📅

File: test/FunctionalTests/DeadlineTests.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            Assert.AreEqual("application/grpc", response.Content.Headers.ContentType.MediaType);
+
+            var responseStream = await response.Content.ReadAsStreamAsync().DefaultTimeout();
+            var pipeReader = new StreamPipeReader(new PipeReaderFixStream(responseStream));
+
+            var messageCount = 0;
+
+            var readTask = Task.Run(async() =>
+            {
+                while (true)
```
Comment (JunTaoLuo) [inline test/FunctionalTests/DeadlineTests.cs]: We should not be relying on this running in a loop. We should explicitly set the clock values for the test and trigger the read/write loops if possible. This may be easier to construct in a DuplexStreaming method.

File: test/FunctionalTests/ServerStreamingMethodTests.cs (hunk 7 above)
Comment (JamesNK) [reply test/FunctionalTests/ServerStreamingMethodTests.cs]: You say that but I found a bug by adding having both tests. One of the call handlers wasn't flushing correctly at the end.

File: test/FunctionalTests/DeadlineTests.cs (hunk 8 above)
Comment (JamesNK) [reply test/FunctionalTests/DeadlineTests.cs]: This will run in a loop for 200 milliseconds and then stop. If it doesn't stop then DefaultTimeout will kill it and throw an error. I don't want to be exact about timings in functional tests because it results in flaky tests.

File: test/FunctionalTests/DeadlineTests.cs (hunk 5 above)
Comment (JamesNK) [reply test/FunctionalTests/DeadlineTests.cs]: This tests the deadline being exceeded.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 4 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

Timeout → "grpc-timeout" TimeoutValue TimeoutUnit
TimeoutValue → {positive integer as ASCII string of at most 8 digits}
TimeoutUnit → Hour / Minute / Second / Millisecond / Microsecond / Nanosecond
Hour → "H"
Minute → "M"
Second → "S"
Millisecond → "m"
Microsecond → "u"
Nanosecond → "n"

File: src/Grpc.AspNetCore.Server/Internal/ClientStreamingServerCallHandler.cs [hunk 9]
```diff
@@ -53,10 +53,16 @@ public override async Task HandleCallAsync(HttpContext httpContext)
...
+            serverCallContext.Initialize();
+
+            TResponse response;
+            using (serverCallContext)
+            {
+                response = await _invoker(
+                    service,
+                    new HttpContextStreamReader<TRequest>(httpContext, Method.RequestMarshaller.Deserializer),
+                    serverCallContext);
+            }
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/ClientStreamingServerCallHandler.cs]: This is fine, though I could also imagine allowing cancellation up until the point where we are done serializing the response and are about the write to the body.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 10]
```diff
@@ -130,5 +133,32 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
...
+                    throw new InvalidOperationException("A timeout greater than 2147483647 milliseconds is not supported.");
+                }
+
+                _deadline = Clock.UtcNow.Add(timeout);
+                _cts = new CancellationTokenSource(timeout);
+            }
+            else
+            {
+                _deadline = DateTime.MaxValue;
+                _cts = new CancellationTokenSource();
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: in this case, how would the _cts be cancelled?

Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Client could cancel.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 1 above)
Comment (jtattermusch) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: the server should send RST_STREAM(CANCEL) if the deadline is exceeded.

The processing on client side should look like this (irrelevant now but will be important for managed client):
When the client receives RST_STREAM, it should double-check its local timer to see if it has expired (it's possible it's expired, but not been processed yet). If it has expired, it should throw away the server "response" (the RST) and generate a DEADLINE_EXCEEDED instead.

File: src/Grpc.AspNetCore.Server/Internal/GrpcProtocolConstants.cs [hunk 11]
```diff
@@ -20,6 +20,7 @@ namespace Grpc.AspNetCore.Server.Internal
 {
     internal static class GrpcProtocolConstants
     {
+        internal const string TimeoutHeader = "grpc-timeout";
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/GrpcProtocolConstants.cs]: Something we should probably address at some point:
Not sure if we are excluding these headers from metadata exposed via ServerCallContext, but in general the grpc-specific headers that have special handling in the framework shouldn't be exposed by visible in metadata (Grpc.Core doesn't expose them because the c-core also doesn't expose them).

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: > the server should send RST_STREAM(CANCEL) if the deadline is exceeded.

What about if the client cancels? What does a client cancellation look like?

I've been struggling to find good docs around gRPC cancellation.

Comment (jtattermusch) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: if client cancels (for whatever reason - it can be that the deadline exceeded on the client side or that the user has decided to cancel), it just sends RST_STREAM(cancel) to notify the server. I'm assuming this is the default behavior when HttpClient decides to cancel an inflight request.
